### ISSUE TITLE
ZStream.groupedWithin works in unexpected way for infinite stream

### ISSUE THREAD

**Issue by DenisKusakin** [zio/zio#8686]
The behavior is only reproducible on **ZIO v1.0.18**, works as expected for elder versions

```scala
import zio.clock.Clock
import zio.duration.durationInt
import zio.stream.ZStream
import zio.{ExitCode, URIO}
object Main extends zio.App {

  override def run(args: List[String]): URIO[zio.ZEnv, ExitCode] = {
    ZStream
      .fromIterable(Seq(1, 2))
      .forever
      .mapAccum(Set[Int]())(
        (acc, i) =>
          if (acc.contains(i)) {
            (acc, None)
          } else {
            (acc + i, Some(i))
        }
      )
      .collectSome
      //The stream above generates only two elements
      // so after 5 seconds those two elements should be grouped
      .groupedWithin(3, 5.seconds)
      .take(1)
      .runCollect
      .map(_ => ExitCode.success)
      .provideCustomLayer(Clock.live)
  }
}
```

**Expected result:** Complete with exit code 0
**Actual result:** The program is never completed

**Comment by krsvital** [comment]
forever ???

**Comment by DenisKusakin** [comment]
> forever ???

Yes, the top-level stream is an infinite stream of 1, 2, 1, 2,..., then there is filtering that produces only two elements: 1, 2. I expect that **groupedWithin** will produce chunk [1, 2] in 5 seconds, **take(1)** will take this chunk and program completed

**Comment by jdegoes** [comment]
We will take a PR for ZIO 1.x issues, of course, but we will need a volunteer to fix this issue.

**Comment by DenisKusakin** [comment]
/attempt #8686
Looks like the issue is related with [PR](https://github.com/zio/zio/pull/7396). I believe that by making concat faster it somehow introduced this behavior

<div id="algora-attempt" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/clvy59w8u0002l209rdh0thw6/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by SAVAMforWIN** [comment]
/attempt #8686

Plan: reproduce the groupedWithin behavior on the still-open issue state, determine whether the remaining bug is in concat/groupedWithin semantics for infinite streams or whether the rewarded earlier fixes left a narrower edge case, and then submit the smallest regression fix with deterministic coverage instead of widening stream internals unnecessarily.

**Comment by Iineman2** [comment]
/attempt #8686

Plan: reproduce the groupedWithin infinite-stream timeout behavior against current series/2.x first, then submit only a focused regression fix if the behavior still exists. If current main has already healed it, I will not open a duplicate PR and will report that evidence instead.

**Comment by artoshirei** [comment]
/attempt #8686

Plan: submit the already-tested series/1.x regression fix for groupedWithin on infinite streams. The patch adds a focused test that times out on baseline and updates aggregateAsyncWithinEither so a fired schedule flushes pending transducer state even when upstream keeps producing empty chunks. I will open this as a draft PR with the test evidence and claim text.

**Comment by Iineman2** [comment]
Update after reproducing locally: on current series/2.x I can add the #8686 live-clock regression and it passes without any production-code change. Focused run: streamsTestsJVM/testOnly *.ZStreamSpec -- -t schedule, including groupedWithin with Seq(1,2) ++ forever-empty, completed successfully.

I am not opening a duplicate/test-only bounty PR from this attempt because the reported behavior does not reproduce on current series/2.x for me. If maintainers want this coverage anyway, I can submit it separately, but I do not want to claim the bounty without a real fix.
